File: android/app/src/main/java/enzio/android/eatfit/data/remote/GrpcDiaryService.kt
package enzio.android.eatfit.data.remote

import enzio.android.eatfit.proto.*
import io.grpc.ManagedChannel
import io.grpc.okhttp.OkHttpChannelBuilder
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import java.util.concurrent.TimeUnit

class GrpcDiaryService(
    private val serverUrl: String = "192.168.0.10",
    private val port: Int = 50051
) {
    private val channel: ManagedChannel = OkHttpChannelBuilder
        .forAddress(serverUrl, port)
        .usePlaintext()
        .keepAliveTime(30, TimeUnit.SECONDS)
        .build()

    private val stub = DiaryServiceGrpc.newBlockingStub(channel)

    suspend fun listConsumptions(userId: String, days: Int = 30): List<ConsumptionGroup> = withContext(Dispatchers.IO) {
        try {
            val request = ListConsumptionsRequest.newBuilder()
                .setUserId(userId)
                .setLimit(days)
                .build()
            stub.listConsumptions(request).groupsList
        } catch (e: Exception) {
            emptyList()
        }
    }

    suspend fun addConsumption(
        userId: String, mealId: String, mealName: String, amount: Double,
        calories: Double, proteins: Double, fats: Double, carbs: Double, water: Double
    ): Boolean = withContext(Dispatchers.IO) {
        try {
            val request = AddConsumptionRequest.newBuilder()
                .setUserId(userId)
                .setMealId(mealId)
                .setMealName(mealName)
                .setAmount(amount)
                .setCalories(calories)
                .setProteins(proteins)
                .setFats(fats)
                .setCarbs(carbs)
                .setWater(water)
                .build()
            stub.addConsumption(request).success
        } catch (e: Exception) {
            false
        }
    }

    suspend fun deleteConsumption(recordId: String): Boolean = withContext(Dispatchers.IO) {
        try {
            stub.deleteConsumption(DeleteConsumptionRequest.newBuilder().setRecordId(recordId).build()).success
        } catch (e: Exception) {
            false
        }
    }

    fun close() { channel.shutdown() }
}

File: desktop/EatfitDesktop/Services/GrpcDiaryService.cs
using System;
using System.Collections.Generic;
using System.Net.Http;
using System.Threading.Tasks;
using Eatfit.V1;
using Grpc.Net.Client;

namespace EatfitDesktop.Services
{
    public class GrpcDiaryService : IDisposable
    {
        private readonly GrpcChannel _channel;
        private readonly Eatfit.V1.DiaryService.DiaryServiceClient _client;

        public GrpcDiaryService(string serverUrl = "http://localhost:50051")
        {
            var handler = new HttpClientHandler
            {
                ServerCertificateCustomValidationCallback = HttpClientHandler.DangerousAcceptAnyServerCertificateValidator
            };
            _channel = GrpcChannel.ForAddress(serverUrl, new GrpcChannelOptions { HttpHandler = handler });
            _client = new Eatfit.V1.DiaryService.DiaryServiceClient(_channel);
        }

        public async Task<List<ConsumptionGroup>> ListConsumptionsAsync(string userId, int days = 30)
        {
            try
            {
                var response = await _client.ListConsumptionsAsync(new ListConsumptionsRequest
                {
                    UserId = userId,
                    Limit = days
                });
                return new List<ConsumptionGroup>(response.Groups);
            }
            catch { return new(); }
        }

        public async Task AddConsumptionAsync(string userId, string mealId, string mealName,
            double amount, double calories, double proteins, double fats, double carbs, double water)
        {
            try
            {
                await _client.AddConsumptionAsync(new AddConsumptionRequest
                {
                    UserId = userId,
                    MealId = mealId,
                    MealName = mealName,
                    Amount = amount,
                    Calories = calories,
                    Proteins = proteins,
                    Fats = fats,
                    Carbs = carbs,
                    Water = water
                });
            }
            catch { }
        }

        public async Task DeleteConsumptionAsync(string recordId)
        {
            try
            {
                await _client.DeleteConsumptionAsync(new DeleteConsumptionRequest { RecordId = recordId });
            }
            catch { }
        }

        public void Dispose() => _channel?.Dispose();
    }
}